beacon-chain/core/epoch/epoch_processing.go

```diff
@@ -91,55 +70,78 @@ func AttestingBalance(ctx context.Context, state state.ReadOnlyBeaconState, atts
 //	 for index in activation_queue[:get_validator_churn_limit(state)]:
 //	     validator = state.validators[index]
 //	     validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
-func ProcessRegistryUpdates(ctx context.Context, state state.BeaconState) (state.BeaconState, error) {
-	currentEpoch := time.CurrentEpoch(state)
-	vals := state.Validators()
+func ProcessRegistryUpdates(ctx context.Context, st state.BeaconState) (state.BeaconState, error) {
+	currentEpoch := time.CurrentEpoch(st)
 	var err error
 	ejectionBal := params.BeaconConfig().EjectionBalance
-	activationEligibilityEpoch := time.CurrentEpoch(state) + 1
-	for idx, validator := range vals {
-		// Process the validators for activation eligibility.
-		if helpers.IsEligibleForActivationQueue(validator, currentEpoch) {
-			validator.ActivationEligibilityEpoch = activationEligibilityEpoch
-			if err := state.UpdateValidatorAtIndex(primitives.ValidatorIndex(idx), validator); err != nil {
-				return nil, err
-			}
+
+	// To avoid copying the state validator set via st.Validators(), we will perform a read only pass
+	// over the validator set while collecting validator indices where the validator copy is actually
+	// necessary, then we will process these operations.
+	eligibleForActivationQ := make([]primitives.ValidatorIndex, 0)
+	eligibleForActivation := make([]primitives.ValidatorIndex, 0)
+	eligibleForEjection := make([]primitives.ValidatorIndex, 0)
+
+	if err := st.ReadFromEveryValidator(func(idx int, val state.ReadOnlyValidator) error {
+		// Collect validators eligible to enter the activation queue.
+		if helpers.IsEligibleForActivationQueue(val, currentEpoch) {
+			eligibleForActivationQ = append(eligibleForActivationQ, primitives.ValidatorIndex(idx))
```

Comment:
This pattern seems useful generically: we can probably split out a helper that takes a predicate and returns the list of validator indices that satisfies the predicate.  We could have it take a slice of predicates and then return a slice of slices with the val indices (in this case it would be three predicates).  PErhaps a benchmark shows that three loops vs 1 loop with three predicates is more or less the same.

Reply:
Yes, I think that would be a good addition as well. I am thinking something like

```go
func (s BeaconState) FilterValidators(fns []func(ReadOnlyValidator) bool) [][]primitives.ValidatorIndex {
  res := make([]primitives.ValidatorIndex, len(fns))
  for i, v := range s.ReadOnlyValidators {
   for j, fn := range fns {
    if fn(v) {
     res[j] = append(res[j], primitives.ValidatorIndex(i))
    }
   }
  }
 return res
}
```

Can I add it in a follow up PR?